server/app/assets/javascript/map_image.coffee

```diff
@@ -80,15 +117,77 @@ class MapImage
       that.onclick(event.clientX - @offsetLeft + nX, event.clientY - @offsetTop + nY)
 
   loadImage: () ->
-    @img = new Image()
-    @img.src = @tag.attr('data-src')
-    @img.onload = =>
-      @setImage()
-      @onload()
+    @imageUrl = @tag.attr('data-src')
+    @bgImg = new Image()
+    that = @
+    @bgImg.onload = () ->
+      if that.layers.length == 0
+        that.setImage()
+        that.onload()
+      else
+        that.loadLayerImages()
+
+    @frameUrl = @tag.attr('data-frame')
+    @bgLayers = []
+    if @frameUrl
+      @bgUrl = @tag.attr('data-background')
+      that = @
+      $.when(
+        $.getJSON(@frameUrl),
+        $.getJSON(@bgUrl)
+      ).done (frameData, bgData) ->
+        frameData[0].forEach (d) ->
+          that.frames[d.name] = {pos: {x: d.posX, y: d.posY}, width: d.width, height: d.height}
+        that.bgLayers = bgData[0].map (bg) -> bg.imageName
+        that.bgImg.src = that.imageUrl
+    else
+      @bgImg.src = @imageUrl
+
+  loadLayerImages: () ->
+    @layerImgs = []
+    layerCount = 0
+    that = @
+    @layers.forEach (s) ->
+      img = new Image()
+      img.onload = () ->
+        layerCount++
+        if that.layers.length == layerCount
+          that.setImage()
+          that.onload()
+
+      that.layerFrames[s] = {}
+      that.layerImgs[s] = img
+      $.getJSON that.frameUrl, {suffix: s}, (data) ->
+        data.forEach (d) ->
+          that.layerFrames[s][d.name] = {pos: {x: d.posX, y: d.posY}, width: d.width, height: d.height}
+        img.src = that.imageUrl + '?suffix=' + s
 
   setImage: () ->
     @ctx.globalAlpha = 1.0
-    @ctx.drawImage(@img, 0, 0, @img.width, @img.height)
+    that = @
+    if @bgLayers.length != 0
+      @bgLayers.forEach (name) ->
+        frame = that.frames[name]
+        that.ctx.drawImage(that.bgImg, frame.pos.x, frame.pos.y, frame.width, frame.height, 0, 0, frame.width, frame.height)
```

Comment:
__[CoffeeLint]__ Line exceeds maximum allowed length Length is 124, max is 120 ([link](https://sider.review/gh/repos/16994149/pulls/434#issue-4040464))

<sub>posted by Sider</sub>